[GAL-590] Email popover

jakzaizzat opened the pull request:
**Changes**

- Add new variant to our toastr component - `error`
- new Toggle component
- Add `headerAction` props to our modal component
- Check JWT token on email verification (logged in and non-logged in user)
- Email banner on notification list

**Demo**

**Email Banner + Cog icon**

<img width="551" alt="CleanShot 2022-11-11 at 20 44 58@2x" src="https://user-images.githubusercontent.com/4480258/201343204-724721b6-b736-4a91-a7f3-52ca1f196c2f.png">

**New setting modal**

<img width="712" alt="CleanShot 2022-11-11 at 20 45 15@2x" src="https://user-images.githubusercontent.com/4480258/201343243-c87448bb-d668-4a9b-8cd6-202bc8578dbb.png">

**Verify Email**

https://user-images.githubusercontent.com/4480258/201608155-e5fc73fd-ce0f-49b2-8ecc-5e8dfe3d5df1.mp4

tbezman commented on src/components/Email/ManageWalletModalWithEmail.tsx:
```diff
@@ -0,0 +1,193 @@
+  const isEmailUnsubscribedFromAll =
+    !!query?.viewer?.email?.emailNotificationSettings?.unsubscribedFromAll ?? false;
+
+  // Set the initial state of the email notification toggle
+  // so we can instantly toggle it on/off without waiting for the mutation to complete
+  useEffect(() => {
+    setIsEmailNotificationChecked(isEmailNotificationSubscribed);
+  }, [isEmailNotificationSubscribed]);
```
`setState` in a `useEffect` is generally not my favorite thing, since it can accidentally cause infinite loops.

I think an easier pattern here would be to just initialize the `useState` with the `isEmailNotificationSubscribed` value.

Something like
```ts
  const [isEmailNotificationChecked, setIsEmailNotificationChecked] = useState(isEmailNotificationSubscribed);
```

 If we do that, we can just remove this useEffect

jakzaizzat replied:
Thats a good idea @tbezman! I've updated with this pattern. Definitely much easier to maintain!

tbezman commented on src/components/Email/ManageWalletModalWithEmail.tsx:
```diff
@@ -0,0 +1,193 @@
+  }, [isEmailNotificationSubscribed]);
+
+  // If the user already have email attached, toggle the email manager ui
+  const userEmail = query?.viewer?.email?.email;
+  useEffect(() => {
+    setIsShowAddEmail(Boolean(userEmail));
+  }, [userEmail]);
```
same thing here

tbezman replied:
I want to call out that the tradeoff here is that if the Relay store updates, our local state won't update anymore. But I feel strongly that you wouldn't want some change in the Relay store to cause "random" changes to happen to a form you're submitting.

tbezman replied:
if that is what we end up wanting, that's fine, but if we initialize `useState` appropriately, at least we won't have a double render.

jakzaizzat replied:
TIL: Didn't know that the relay store affect the local state too. 

I've removed this `useEffect` and set the response as a initial local state.

tbezman commented on src/components/Email/ManageWalletModalWithEmail.tsx:
```diff
@@ -0,0 +1,193 @@
+
+        // If its failed, revert the toggle state
+        if (!response?.updateEmailNotificationSettings) {
+          // setIsEmailNotificationChecked(!checked);
```
left over comment?

tbezman commented on src/components/Email/ManageWalletModalWithEmail.tsx:
```diff
@@ -0,0 +1,193 @@
+    async (checked: boolean) => {
+      const unsubscribedFromNotifications = !checked;
+      setIsEmailNotificationChecked(checked);
+      setIsPending(true);
```
`usePromisifiedMutation` returns `[submit, boolean]`. can we thread that through here so we don't have to manage that state ourselves?

tbezman commented on src/components/Email/ManageWalletModalWithEmail.tsx:
```diff
@@ -0,0 +1,193 @@
+          message: 'Settings successfully updated. You will now receive notification emails',
+        });
+        return;
+      } catch (error) {
```
can we tell the user something went wrong?

jakzaizzat replied:
Updated it!

tbezman commented on src/components/Email/ManageWalletModalWithEmail.tsx:
```diff
@@ -0,0 +1,193 @@
+    [isEmailUnsubscribedFromAll, pushToast, reportError, updateEmailNotificationSettings]
+  );
+
+  const toggleEmailNotification = useCallback(() => {
+    // Invert the value from subscribe to unsubscribe
+    handleEmailNotificationChange(!isEmailNotificationChecked);
+  }, [handleEmailNotificationChange, isEmailNotificationChecked]);
```
You're passing this callback to `<Toggle onChange>`. That callback seems to pass the new value as a parameter, if we use that parameter instead of relying on `isEmailNotificationChecked`, then `toggleEmailNotification` will be stable and won't cause re-renders.

jakzaizzat replied:
Good idea! Updated to use native checked value instead of `isEmailNotificationChecked`

tbezman commented on src/components/Email/useVerifyEmailOnPage.ts:
```diff
@@ -0,0 +1,97 @@
+  const { verifyEmail } = router.query;
+  const { pushToast } = useToastActions();
+
+  const updater: SelectorStoreUpdater<useVerifyEmailOnPageMutation['response']> = (
+    store,
+    response
+  ) => {
+    if (response.verifyEmail?.__typename === 'VerifyEmailPayload' && response.verifyEmail?.email) {
+      store
+        .get('client:root:viewer')
+        ?.getLinkedRecord('email')
+        ?.setValue('Verified', 'verificationStatus');
+    }
+  };
```
I don't think we need a custom updater here? We should be able to use `optimisticResponse` instead since we're not doing anything too crazy.

`optimisticResponse` should be strongly typed, you can use it like this
```ts
someMutation({
  variables: {},
  optimisticResponse: {
     verifyEmail: {
       // your data here
     }
  }
})
```

As long as you put `@raw_response_type` at the end of your mutation (which you did), the types for the `optimisticResponse` field will be generated for you.

jakzaizzat replied:
@tbezman  Make sense! Since its just a simple updated. 

In this case, we want to update the `verificationStatus` but it doesn't exist inside the `VerifyEmailPayload`

```
type VerifyEmailPayload {
    email: String
}
```

What's your recommend on this? 

1. Should we ask backend to return the verification status, so we can used the `optimisticResponse`
2. Is there any way to update the value of `verificationStatus` from `pending` to `verified` without using `updater`?

The goal is to updated the verification status in this modal (video below):

https://user-images.githubusercontent.com/4480258/201610416-fa50c190-1aa5-4ce4-a109-3e988ad05858.mp4

tbezman commented on src/contexts/globalLayout/GlobalNavbar/ProfileDropdown/ProfileDropdownContent.tsx:
```diff
@@ -56,32 +58,61 @@ export function ProfileDropdownContent({ showDropdown, onClose, queryRef }: Prop
       isFullPage: isMobile,
       isPaddingDisabled: true,
       headerVariant: 'standard',
+      headerActions: isEmailFeatureEnabled ? notificationModalActions : false,
```
instead of making `showModal` take a one off prop, can we modify the one off `NotificationsModal` to show the right stuff? this way, we don't keep bloating the `showModal` function with more options.

jakzaizzat replied:
Make sense! Move it to its own hooks/component.

Robinnnnn commented on src/components/Email/EmailForm.tsx:
```diff
@@ -17,9 +17,10 @@ import { HStack, VStack } from '../core/Spacer/Stack';
 type Props = {
   setIsEditMode: (editMode: boolean) => void;
   queryRef: EmailFormFragment$key;
+  onClosed?: () => void;
```
nit: rename => onClose

Robinnnnn commented on src/components/Email/ManageWalletModalWithEmail.tsx:
```diff
@@ -0,0 +1,193 @@
+  onTezosAddWalletSuccess?: () => void;
+};
+
+function ManageWalletsModalWithEmail({
```
nit: should we just call this `SettingsModal`?

Robinnnnn commented on src/components/Email/ManageWalletModalWithEmail.tsx:
```diff
@@ -0,0 +1,193 @@
+  const { pushToast } = useToastActions();
+  const reportError = useReportError();
+
+  // Invert the value since in the db stored as unsubscribe while in the UI it's subscribe
+  const isEmailNotificationSubscribed =
+    query?.viewer?.email?.emailNotificationSettings?.unsubscribedFromNotifications || false;
+
+  const isEmailUnsubscribedFromAll =
+    !!query?.viewer?.email?.emailNotificationSettings?.unsubscribedFromAll ?? false;
```
hmm i think this can just be the following, without the need for fallbacks, because the `!` will coerce them into booleans:

```js
const isEmailNotificationSubscribed = !query?.viewer?.email?.emailNotificationSettings?.unsubscribedFromNotifications

const isEmailUnsubscribedFromAll = !!query?.viewer?.email?.emailNotificationSettings?.unsubscribedFromAll
```

jakzaizzat replied:
Removed the fallback!

Robinnnnn commented on src/components/Email/ManageWalletModalWithEmail.tsx:
```diff
@@ -0,0 +1,193 @@
+          </HStack>
+        </VStack>
+        <StyledButtonContaienr>
+          {isShowAddEmail ? (
```
can we rename this `shouldDisplayAddEmailInput`

Robinnnnn commented on src/components/NotificationsModal/NotificationEmailAlert.tsx:
```diff
@@ -0,0 +1,81 @@
+
+  const { showModal, hideModal } = useModalActions();
+
+  const handleEnableEmails = () => {
+    // Hide notification modal
+    hideModal();
+
+    showModal({
+      content: <ManageWalletsModal queryRef={query} />,
+      headerText: 'Manage accounts',
+    });
+  };
+  const handleDismiss = () => {
+    onDismiss();
+  };
```
nit: could we wrap these in `useCallback`?

Robinnnnn commented on src/components/core/Toggle/Toggle.tsx:
```diff
@@ -0,0 +1,70 @@
+} & React.InputHTMLAttributes<HTMLInputElement>;
+
+export default function Toggle({ checked, isPending, onChange, ...inputProps }: Props) {
+  const handleChange = (e: React.ChangeEvent<HTMLInputElement>) => {
+    onChange(e.target.checked);
+  };
```
useCallback

Robinnnnn commented on src/contexts/toast/ToastContext.tsx:
```diff
@@ -34,19 +35,23 @@ type ToastType = {
   message: string;
   onDismiss: DismissToastHandler;
   autoClose: boolean;
+  variant: 'success' | 'error';
```
nit: looks like this is repeated in Line 11

could do `type ToastVariant = 'success' | 'error'` and use that

Robinnnnn reviewed (commented):
amazing work jakz!!

1. in the video, it looks like the modal "stutters" a little bit – does this happen regularly?

2. the change from "Pending" to "Verified" looks like it causes some layout shift, and the height of the section decreases a bit – can we make sure the height remains consistent during verification?

![image](https://user-images.githubusercontent.com/12162433/201444171-64ef4326-07df-4f06-9e7d-a8a62bb6b01f.png)

![image](https://user-images.githubusercontent.com/12162433/201444180-51a2cc09-9f18-4df6-b4ca-eaea867914b0.png)

jakzaizzat commented:
Thanks @Robinnnnn! 

1. Regarding the slutters modal, yes its happens 🤦 . I've updated to make sure its open the modal first and then load the user gallery. 

https://user-images.githubusercontent.com/4480258/201608433-aba37767-19c5-4724-8320-9cd11391f0f8.mp4

2. Layout shift: Updated it!
